aten/src/ATen/core/dispatch/OperatorEntry.cpp

```diff
@@ -153,49 +153,71 @@ const KernelFunction& OperatorEntry::computeDispatchTableEntry(const c10::Dispat
   //          if there's no direct registration to the backend key.
   //          Tensor factory functions used to have no registration to Autograd key but only to catchAll.
   //          In the past we directly call into backends(filled with catchAll) after BackendSelect.
   //          Now that we first call Autograd backend keys after BackendSelect, we should fill those
   //          with catchAll as well.
   //  (3) Use fallthrough kernel that are registered as fallback.
   //  (4) Use catchAll kernel if available
-  // TODO: currently Autograd is the only alias key, we'll update alias key precedence after we add new
-  //      alias keys AutogradDispatchCPUOrCUDA and Math.
-  // TODO: we can fix (2.2) and remove (4) after TypeDefault registrations are moved from catchAll to Math
+  // Alias Key Precedence:
+  //   Math > Autograd
+  // TODO: Update alias key precedence after we add new alias keys AutogradDispatchCPUOrCUDA .
+  // TODO: we can remove (2.3) and (4) after TypeDefault registrations are moved from catchAll to Math
   //       so that Math can populate to Autograd backend keys before fallback kernels.
 
   // 1. Operator registration
-  auto kern_it = kernels_.find(dispatch_key);
-  if (kern_it != kernels_.end()) {
-    TORCH_INTERNAL_ASSERT(!kern_it->second.empty());
-    TORCH_INTERNAL_ASSERT(kern_it->second.front().kernel.isValid());
-    return {kern_it->second.front(), "kernel"};
-
-  } else if (isIncludedInAlias(dispatch_key, DispatchKey::Autograd)) {
-    // 2.1. For autograd backend keys, use kernel from DispatchKey::Autograd if available
-    auto kern_autograd = kernels_.find(DispatchKey::Autograd);
-    if (kern_autograd != kernels_.end()) {
-      TORCH_INTERNAL_ASSERT(!kern_autograd->second.empty());
-      TORCH_INTERNAL_ASSERT(kern_autograd->second.front().kernel.isValid());
-      return {kern_autograd->second.front(), "autograd kernel"};
-
-    // 2.2. For autograd backend keys, we use kernel from alias Math key (catchAll will be moved to Math)
-    //      if there's no direct registration to the backend key.
-    } else if (kernels_.find(getBackendKeyFromAutograd(dispatch_key)) == kernels_.end()
-            && !catchAllKernel_.empty()) {
-      TORCH_INTERNAL_ASSERT(catchAllKernel_.front().kernel.isValid());
-      return {catchAllKernel_.front(), "catch all"};
-    }
+  if (hasKernelForDispatchKey(dispatch_key)) {
+    return {getKernelForDispatchKey(dispatch_key), "kernel"};
+  }
+
+  // 2.1. Use Math kernel if available. For autograd keys, we only use kernel from Math
+  //      when there's no direct registration to its corresponding backend key.
+  if (isIncludedInAlias(dispatch_key, DispatchKey::Math)
+      && hasKernelForDispatchKey(DispatchKey::Math)
+      && !(isIncludedInAlias(dispatch_key, DispatchKey::Autograd)
+           && hasKernelForDispatchKey(getBackendKeyFromAutograd(dispatch_key)))) {
```

Comment:
The "backend that uses AutogradOther" case will be missed here, bc `getBackendKeyFromAutograd(AutogradOther)` returns undefined - I think this is a problem. 

Most immediately, since the data model says there's a 1:n relationship between autograd keys and backend keys, I think the correct API for moving along this mapping is `getBackendKeySetFromAutograd(k)` that returns a keyset of backends in all cases (though ofc it might be a singleton). (Apologies that I didn't get a chance to review the PR where this went in!)

Once we have that, then the question becomes how to generalize the "only use Math if no backend kernel has been registered" condition to sets of backends. Unfortunately I think any simple answer - say, anding the "no kernel" constraint across the set - reintroduces the problem of nonlocal effects of a registration change. 

Going to leave this comment here as-is before thinking more about it, just in case I've made a basic mistake and it's not really a problem :) otherwise, let's chat about it at tomorrow morning's thing.

Reply:
Great point! It's true that current implementation will always pick up `Math` kernel if exists. Maybe a workaround could be `make Autograd higher precedence than Math for AutogradOther`? 
Happy to followup on this tmr morning :D

Reply:
There is no right answer here. If it truly is a show stopper, the right thing to do is to move the key in question out of AutogradOther. I don't mind either default, but Math is probably easier for these exotic custom backends (e.g., the Math version implies less work for a custom backend; if they want to do more work, then they can also ask us to split out their key).